dateutil/parser.py

```diff
@@ -50,15 +50,20 @@ class _timelex(object):
     _split_decimal = re.compile("([.,])")
 
     def __init__(self, instream):
-        if isinstance(instream, binary_type):
-            instream = instream.decode()
-
-        if isinstance(instream, text_type):
-            instream = StringIO(instream)
-
-        if getattr(instream, 'read', None) is None:
-            raise TypeError('Parser must be a string or character stream, not '
-                            '{itype}'.format(itype=instream.__class__.__name__))
+        if not hasattr(instream, 'read'):
```

Comment:
I do prefer the `getattr(instream, 'read', None) is None` idiom over `hasattr`, at least until Python 2.7 support is dropped, based on [this article](https://hynek.me/articles/hasattr/).

Also, why is this being checked at all? Just to avoid streams that have a `.decode()` method?